### PR TITLE
Trait Points

### PR DESCRIPTION
**Pull request by DEATHB4DEFEAT**
# Description

Some improvements to loadouts too.

---

# TODO

- [x] Points logic
- [x] Server-side validation
- [x] Categorize traits
- [x] Assign points to traits
- [x] Header costs
- [x] Sort entries
- [x] Max traits
	- [x] Communicate max traits
- [x] Point bar
- [x] Group exclusivity
- Black outline on text
- [x] Fix existing component whitelists

---

<details><summary><h1>Media</h1></summary>
<p>

## Accurate except for small details

image

### Something to note:

image

image

image

</p>
</details>

---

# Changelog

:cl:
- add: Added trait points
- add: Added categories for traits

### CODE REVIEW

File: Content.Shared/CCVar/CCVars.cs [hunk 1]
```diff
@@ -341,14 +341,34 @@ public sealed class CCVars : CVars
...
+        ///     Whether to allow characters to select traits.
+        /// </summary>
+        public static readonly CVarDef<bool> GameTraitsEnabled =
+            CVarDef.Create("game.traits_enabled", true, CVar.REPLICATED);
+
+        /// <summary>
+        ///     How many traits a character can have at most.
+        /// </summary>
+        public static readonly CVarDef<int> GameTraitsMax =
+            CVarDef.Create("game.traits_max", Int32.MaxValue, CVar.REPLICATED);
```
Comment (DangerRevolution) [inline Content.Shared/CCVar/CCVars.cs:355]: It's really late; sorry if I'm staring at it and missing it but where is the value for how many traits can be selected maximum?

**Review by DangerRevolution — APPROVED** [review]
Approved with just one question

File: Content.Shared/CCVar/CCVars.cs (hunk 1 above)
Comment (DEATHB4DEFEAT) [reply Content.Shared/CCVar/CCVars.cs:355]: `Int32.MaxValue`, a practically infinite number.

Comment (DangerRevolution) [reply Content.Shared/CCVar/CCVars.cs:355]: so if I wanted to set it to 5 traits maximum; how would I change this?

**Review by VMSolidus — APPROVED** [review]
I'm satisfied with this system now. My only complaint is something out of scope of this PR, but is something a ton of people are going to solve pretty shortly once we've merged this, myself included. That is, the game outright lacks positive traits for people to spend their trait points on. The entire list is all negative or neutral traits.

**Comment by DangerRevolution** [comment]
> DEATHB4DEFEAT added Status: Do Not Merge Status: Help Wanted labels 43 minutes ago

**Comment by VMSolidus** [comment]
> > DEATHB4DEFEAT added Status: Do Not Merge Status: Help Wanted labels 43 minutes ago

Yea I was with him in a voice call doing what he needed done. :)

**Comment by DangerRevolution** [comment]
> > > DEATHB4DEFEAT added Status: Do Not Merge Status: Help Wanted labels 43 minutes ago
> 
> Yea I was with him in a voice call doing what he needed done. :)

Labels are a lie.. 1984

**Pull request merged**